src/comm/aca_grpc.cpp

```diff
@@ -38,124 +38,124 @@ extern string g_ncm_port;
-    if (rc == EXIT_SUCCESS) {
-      ACA_LOG_INFO("Control Fast Path streaming - Successfully updated host with latest goal state %d.\n",
-                   rc);
-    } else if (rc == EINPROGRESS) {
-      ACA_LOG_INFO("Control Fast Path streaming - Update host with latest goal state returned pending, rc=%d.\n",
-                   rc);
-    } else {
-      ACA_LOG_ERROR("Control Fast Path streaming - Failed to update host with latest goal state, rc=%d.\n",
-                    rc);
+  void* got_tag;
+  bool ok;
+  new GoalStateProvisionerAsyncInstance(&service_, cq_.get());
+  while (true)
+  { 
+    if (!cq_->Next(&got_tag, &ok)) {
+        ACA_LOG_DEBUG("Completion Queue Shut. Quitting\n");
+        break;
     }
-    std::chrono::_V2::steady_clock::time_point end = std::chrono::steady_clock::now();
-    auto message_total_operation_time =
-            std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
-
-    ACA_LOG_INFO("[METRICS] Received goalstate at: [%ld], update finished at: [%ld]\nElapsed time for update goalstate operation took: %ld microseconds or %ld milliseconds\n",
-                 start, end, message_total_operation_time,
-                 (message_total_operation_time / 1000));
-    stream->Write(gsOperationReply);
-    gsOperationReply.Clear();
+    static_cast<GoalStateProvisionerAsyncInstance*>(got_tag)->PushGoalStatesStream(ok);
   }
-
-  return Status::OK;
-}
-
-Status GoalStateProvisionerImpl::ShutDownServer()
-{
-  ACA_LOG_INFO("%s", "Shutdown server");
-  server->Shutdown();
-  return Status::OK;
 }
 
-void GoalStateProvisionerImpl::ConnectToNCM()
+void GoalStateProvisionerAsyncServer::RunServer(int thread_pool_size)
 {
-  ACA_LOG_INFO("%s\n", "Trying to init a new sub to connect to the NCM");
-  grpc::ChannelArguments args;
-  // Channel does a keep alive ping every 10 seconds;
-  args.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, 10000);
-  // If the channel does receive the keep alive ping result in 20 seconds, it closes the connection
-  args.SetInt(GRPC_ARG_KEEPALIVE_TIMEOUT_MS, 20000);
-
-  // Allow keep alive ping even if there are no calls in flight
-  args.SetInt(GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS, 1);
+  ServerBuilder builder;
+  string GRPC_SERVER_ADDRESS = "0.0.0.0:" + g_grpc_server_port;
+  builder.AddListeningPort(GRPC_SERVER_ADDRESS, grpc::InsecureServerCredentials());
+  builder.RegisterService(&service_);
+  cq_ = builder.AddCompletionQueue();
+  server_ = builder.BuildAndStart();
+  ACA_LOG_INFO("Async GRPC: Streaming capable GRPC server listening on %s\n",
+               GRPC_SERVER_ADDRESS.c_str());
 
-  chan_ = grpc::CreateCustomChannel(g_ncm_address + ":" + g_ncm_port,
-                                    grpc::InsecureChannelCredentials(), args);
-  stub_ = GoalStateProvisioner::NewStub(chan_);
+  thread_pool_.resize(thread_pool_size);
+  std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
is sleep needed here, resize in _aca_on_demand_engine.h_ doesn't have it

Reply:
@tianyuan129 , if we don't need it here, please delete it.

Reply:
If we don't wait here, later when try to push the gRPC server into the thread pool, it will see 0 idle threads in the pool. My guess is that it needs time to initialize the thread pool. `aca_on_demand_engine.h` doesn't have it because the task is pushed into the pool much mater after the pool initialized.

Reply:
make sense, we can leave the sleeping there for now but later we can figure out a smarter way to achieve the same thing

Reply:
Fixed